src/dataflow/src/render/filter.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use differential_dataflow::lattice::Lattice;
+use differential_dataflow::Collection;
+
+use timely::dataflow::Scope;
+use timely::progress::{timestamp::Refines, Timestamp};
+
+use dataflow_types::DataflowError;
+use expr::{RelationExpr, ScalarExpr};
+use repr::{Datum, Row};
+
+use crate::operator::CollectionExt;
+use crate::render::context::Context;
+
+impl<G, T> Context<G, RelationExpr, Row, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+{
+    /// Finds collection corresponding to input RelationExpr and then applies
+    /// predicates to the input collection.
+    pub fn render_filter(
+        &mut self,
+        input: &RelationExpr,
+        predicates: Vec<ScalarExpr>,
+    ) -> (Collection<G, Row>, Collection<G, DataflowError>) {
+        let (ok_collection, err_collection) = self.collection(input).unwrap();
```

Comment:
This now has the slightly awkward issue that the `unwrap` call does not follow an `ensure_rendered` call. Since the error isn't as clearly localized, we should probably put a message in there (or find a different way to ensure that we expect to find the relation expr). It is also worth documenting that the method will panic if `input` hasn't yet been rendered (or isn't otherwise available in `self`).

Reply:
Or, we could just call `ensure_rendered` here, I think? Is there a reason to separate that out?